internal/controller/healthcheck/healthcheck_controller.go

```diff
@@ -0,0 +1,203 @@
+	Client client.Client
+}
+
+const (
+	providerLabelKey = "cluster.x-k8s.io/provider"
+)
+
+func (r *ProviderHealthCheckReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.Deployment{}, builder.WithPredicates(providerDeploymentPredicates())).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *ProviderHealthCheckReconciler) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.Info("Checking provider health")
+
+	result := ctrl.Result{}
+
+	deployment := &appsv1.Deployment{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, deployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return result, nil
+		}
+		// Error reading the object - requeue the request.
+		return result, err
+	}
+
+	// There should be just one owner reference - to a Provider resource.
+	if len(deployment.GetOwnerReferences()) != 1 {
+		return result, fmt.Errorf("incorrect number of owner references for provider deployment %s", req.NamespacedName)
+	}
+
+	deploymentOwner := deployment.GetOwnerReferences()[0]
+
+	deploymentCondition := getDeploymentCondition(deployment.Status, appsv1.DeploymentAvailable)
+
+	typedProvider, err := r.getGenericProvider(ctx, deploymentOwner.Kind, deploymentOwner.Name, req.Namespace)
+	if err != nil {
+		return result, err
+	}
+
+	// Stop earlier if this provider is not fully installed yet.
+	if !conditions.IsTrue(typedProvider, operatorv1.ProviderInstalledCondition) {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	// Compare provider's Ready condition with the deployment's Available condition and stop if they already match.
+	currentReadyCondition := conditions.Get(typedProvider, clusterv1.ReadyCondition)
+	if currentReadyCondition != nil && deploymentCondition != nil && currentReadyCondition.Status == deploymentCondition.Status {
+		return result, nil
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(typedProvider.GetObject(), r.Client)
+	if err != nil {
+		return result, err
+	}
+
+	if deploymentCondition != nil {
+		conditions.Set(typedProvider, &clusterv1.Condition{
+			Type:   clusterv1.ReadyCondition,
+			Status: deploymentCondition.Status,
+			Reason: deploymentCondition.Reason,
+		})
+	} else {
+		conditions.Set(typedProvider, &clusterv1.Condition{
+			Type:   clusterv1.ReadyCondition,
+			Status: corev1.ConditionFalse,
+		})
+	}
+
+	// Don't requeue immediately if the deployment is not ready, but rather wait 5 seconds.
+	if conditions.IsFalse(typedProvider, clusterv1.ReadyCondition) {
+		result = ctrl.Result{RequeueAfter: 5 * time.Second}
+	}
+
+	options := patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{clusterv1.ReadyCondition}}
```

Comment:
Yeah, but since the spec stays the same, it stops almost immediately https://github.com/kubernetes-sigs/cluster-api-operator/blob/main/internal/controller/genericprovider_controller.go#L121-L125